src/sys/socket/mod.rs

```diff
@@ -2002,7 +2002,7 @@ unsafe fn pack_mhdr_to_receive<S>(
     let mut mhdr = mem::MaybeUninit::<msghdr>::zeroed();
     let p = mhdr.as_mut_ptr();
     unsafe {
-        (*p).msg_name = address as *mut c_void;
+        (*p).msg_name = if S::size() == 0 { ptr::null_mut() } else { (*address).as_mut_ptr() as *mut c_void };
```

Comment:
> (*address).as_mut_ptr() as *mut c_void

Any idea why it would work after reverting the change made in #2041?

Reply:
> [not sure if things fixed by https://github.com/nix-rust/nix/pull/2041 are still working after that.](https://github.com/nix-rust/nix/issues/2506#issuecomment-2366939632)

Considering this, gentle ping on @JarredAllen, the author of #2041, could you please take a look at this PR?

## Context 

@pacak is trying to solve #2506, after digging into the previous Nix versions, [he figured out that the issue is possibly introduced by #2041](https://github.com/nix-rust/nix/issues/2506#issuecomment-2366915152):

<img width="1652" alt="Screenshot 2024-10-31 at 8 33 28 AM" src="https://github.com/user-attachments/assets/6188eee9-3af7-48b0-adb2-afeeca78db82">

Reply:
> Any idea why it would work after reverting the change made in https://github.com/nix-rust/nix/pull/2041?

As far as I understand `.as_mut_ptr()` gives a null pointer in case of  `()`, `address as *mut c_void` gives a dangling pointer through magic of rustc optimizations.

Reply:
I guess just using `.as_mut_ptr()` would be enough as long as all the instances return the right pointer. Explicit checking for size makes it so it works even if there's a bogus instance.

Reply:
> As far as I understand .as_mut_ptr() gives a null pointer in case of (), address as *mut c_void gives a dangling pointer through magic of rustc optimizations.

If so

> I guess just using .as_mut_ptr() would be enough as long as all the instances return the right pointer.

then yes


-----

> not sure if things fixed by https://github.com/nix-rust/nix/pull/2041 are still working after that.

I think this PR won't break the fix done by #2041 as there is a unit test guarding it.

BTW, let's document why we should use `.as_mut_tr()` rather than `address as *mut c_void` because it is so easy to revert the fix if one is not aware of such a **subtle** different behavior . 😮‍💨

Reply:
Sorry I wasn't around, but thanks for fixing my mistake! I concur with the assessment that this shouldn't break my fix.

Reply:
No need to be sorry. 😆